packages/chain-mon/src/faultproof-wd-mon/service.ts

```diff
@@ -0,0 +1,519 @@
+          rootClaim,
+          l2BlockNumber
+        )
+        if (isValidRoot) {
+          // Check if the withdrawal exists on L2.
+          const exists = await this.state.messenger.sentMessages(hash)
+          // Hopefully the withdrawal exists!
+          if (exists) {
+            // Unlike below we don't grab the timestamp here because it adds an unnecessary request.
+            this.logger.info(`valid withdrawal`, {
+              withdrawalHash: event.args.withdrawalHash,
+            })
+
+            // Bump the withdrawals metric so we can keep track.
+            this.metrics.withdrawalsValidated.inc()
+          } else {
+            this.state.invalidProofWithdrawals.push(disputeGameData)
+            // Grab and format the timestamp so it's clear how much time is left.
+            const block = await event.getBlock()
+            const ts = `${dateformat(
+              new Date(block.timestamp * 1000),
+              'mmmm dS, yyyy, h:MM:ss TT',
+              true
+            )} UTC`
+
+            // Uh oh!
+            this.logger.error(`withdrawalHash not seen on L2`, {
+              withdrawalHash: event.args.withdrawalHash,
+              provenAt: ts,
+            })
+
+            // Change to forgery state.
+            this.state.forgeryDetected = true
+            this.metrics.isDetectingForgeries.set(
+              Number(this.state.forgeryDetected)
+            )
+          }
+        } else {
+          this.state.invalidProposalWithdrawals.push(disputeGameData)
+          this.logger.info(`invalid proposal`, {
+            withdrawalHash: event.args.withdrawalHash,
+          })
+        }
+      }
+    }
+
+    // Increment the highest unchecked block number for the next loop.
+    this.state.highestUncheckedBlockNumber = toBlockNumber
+  }
+
+  /**
+   * Retrieves the dispute games data associated with a withdrawal hash associated in an event.
+   *
+   * @param event The event containing the withdrawal hash.
+   * @returns An array of objects containing the withdrawal hash, sender address, and dispute game address.
+   */
+  async getWithdrawalDisputeGames(event: ethers.Event): Promise<
+    Array<{
+      withdrawalHash: string
+      senderAddress: string
+      disputeGame: ethers.Contract
+    }>
+  > {
+    const withdrawalHash = event.args.withdrawalHash
+    const disputeGameMap: Array<{
+      withdrawalHash: string
+      senderAddress: string
+      disputeGame: ethers.Contract
+    }> = []
+
+    const numProofSubmitter = await this.state.portal.numProofSubmitters(
+      withdrawalHash
+    )
+
+    // iterate for numProofSubmitter
+    const proofSubmitterAddresses = await Promise.all(
+      Array.from({ length: numProofSubmitter.toNumber() }, (_, i) =>
+        this.state.portal.proofSubmitters(withdrawalHash, i)
+      )
+    )
+
+    // Iterate for proofSubmitterAddresses and query provenWithdrawals to get the disputeGameProxy for each proofSubmitter
+    await Promise.all(
```

Comment:
Likely fine but noting for future reference may need to break this up to avoid rate limiting. I think it should be OK but maybe worth commenting.